feature: V4 Post policies

IlyaFaer opened the pull request:
Towards #8

IlyaFaer commented:
@frankyn, I've pushed a real PR. Only system test are failing with `Anonymous caller does not have storage.objects.create access to <bucket-name>/<file-name>` (the same Jonathan Lui got)

Everything except this seems to be fine

frankyn commented:
Thanks @IlyaFaer, I'll review. I'll update group in V4 Sig Hangouts Chat.

IlyaFaer commented on google/cloud/storage/client.py:
```diff
@@ -836,6 +841,106 @@ def get_hmac_key_metadata(
+        bucket_bound_hostname=None,
+        scheme=None,
+    ):
+        """Generate a V4 signed policy object.
```
Added

frankyn commented on google/cloud/storage/_signing.py:
```diff
@@ -629,6 +627,18 @@ def generate_signed_url_v4(
     )
 
 
+def get_v4_dtstamps():
```
Thanks for clarifying.

Name suggestion: `get_v4_now_dtstamps()`

frankyn commented on google/cloud/storage/client.py:
```diff
@@ -836,6 +841,126 @@ def get_hmac_key_metadata(
         metadata.reload(timeout=timeout)  # raises NotFound for missing key
         return metadata
 
+    def get_signed_policy_v4(
```
Pending final confirmation on naming, leaving this as an open item.

IlyaFaer replied:
Done, working now on conformance tests

frankyn commented on google/cloud/storage/client.py:
```diff
@@ -836,6 +841,126 @@ def get_hmac_key_metadata(
+            )
+            >>> with open("bucket-name", "rb") as f:
+                files = {"file": ("bucket-name", f)}
+                requests.post(policy["url"], data=policy["fields"], files=files)
```
Great experience.

frankyn commented on google/cloud/storage/client.py:
```diff
@@ -836,6 +841,126 @@ def get_hmac_key_metadata(
+        fields=None,
+        virtual_hosted_style=False,
+        bucket_bound_hostname=None,
+        scheme=None,
```
Missing some support for additional credential flows such as SignBlob API introduced into signedUrls in #9889.

IlyaFaer replied:
Done

IlyaFaer commented:
@frankyn, @crwilcox, I've pushed conformance tests into separate branch and compared it against this one (to make changes more visual). Here it is: https://github.com/q-logic/python-storage/pull/9

I've already tried to run them with test data manually inserted into `url_signer_v4_test_data.json` - works fine. Will push test data into conformance tests repo shortly.

frankyn commented on google/cloud/storage/client.py:
```diff
@@ -836,6 +846,165 @@ def get_hmac_key_metadata(
 
+    def generate_signed_post_policy_v4(
+        self,
+        credentials,
```
Right now credentials is a positional argument and should be optional like generate_signed_url -> https://github.com/googleapis/python-storage/blob/a50cdd145d39cb67e0d7021c376acbe376436178/google/cloud/storage/blob.py#L363

IlyaFaer replied:
Done

IlyaFaer commented:
@frankyn, pushed everything that I have for now.
For conformance tests see the [second PR](https://github.com/q-logic/python-storage/pull/9). All of them passed with OK.

Got `lint` session failed in kokoro, but I'm using the same version of `black` on my computer than kokoro does, and it says that file left unchanged.

crwilcox commented on google/cloud/storage/client.py:
```diff
@@ -836,6 +846,168 @@ def get_hmac_key_metadata(
+                "bucket-name",
+                "blob-name",
+                conditions=[
+                    ['content-length-range', 0, 255]
```
Could you modify this to use `"` and not `'`?

crwilcox commented on tests/unit/test_client.py:
```diff
@@ -1471,3 +1485,266 @@ def test_get_hmac_key_metadata_w_project(self):
+        self.assertEqual(fields["x-goog-signature"], "0c4003044105")
+        self.assertEqual(
+            fields["policy"],
+            b"eyJjb25kaXRpb25zIjpbeyJidWNrZXQiOiJidWNrZXQtbmFtZSJ9LHsiYWNsIjoicHJpdmF0ZSJ9LFsic3RhcnRzLXdpdGgiLCIkQ29udGVudC1UeXBlIiwidGV4dC9wbGFpbiJdLHsia2V5Ijoib2JqZWN0LW5hbWUifSx7IngtZ29vZy1kYXRlIjoiMjAyMDAzMTJUMTE0NzE2WiJ9LHsieC1nb29nLWNyZWRlbnRpYWwiOiJ0ZXN0QG1haWwuY29tLzIwMjAwMzEyL2F1dG8vc3RvcmFnZS9nb29nNF9yZXF1ZXN0In0seyJ4LWdvb2ctYWxnb3JpdGhtIjoiR09PRzQtUlNBLVNIQTI1NiJ9XSwiZXhwaXJhdGlvbiI6I ...
```
Can we extract this to a const?

frankyn commented on google/cloud/storage/client.py:
```diff
@@ -836,6 +846,168 @@ def get_hmac_key_metadata(
+                fields=[
+                    'x-goog-meta-hello' => 'world'
+                ],
+                expiration=datetime.datetime(2020, 3, 17),
```
move expiration before conditions*

frankyn commented on google/cloud/storage/client.py:
```diff
@@ -836,6 +846,168 @@ def get_hmac_key_metadata(
+            signature = _sign_message(str_to_sign, access_token, service_account_email)
+            signature_bytes = base64.b64decode(signature)
+        else:
+            signature_bytes = credentials.sign_bytes(str_to_sign)
```
Add commentary about what this code is doing.

frankyn commented on google/cloud/storage/client.py:
```diff
@@ -836,6 +846,168 @@ def get_hmac_key_metadata(
+            url = "https://{}.storage.googleapis.com/".format(bucket_name)
+
+        elif bucket_bound_hostname:
+            if ":" in bucket_bound_hostname:
```
Add a comment to explain this conditional check.

frankyn commented on google/cloud/storage/client.py:
```diff
@@ -836,6 +846,168 @@ def get_hmac_key_metadata(
+
+            >>> from google.cloud import storage
+            >>> client = storage.Client()
+            >>> client.generate_signed_post_policy_v4(
```
Missing `policy = `

IlyaFaer commented:
@frankyn, @crwilcox, I've moved conformance tests into this PR. I assume you'd like me to add the conformance tests data as well!? Locally I've added policy conformance tests data into `url_signer_v4_test_data.json`, and used it while running the tests. Is it a correct place?

![Безымянный](https://user-images.githubusercontent.com/26030071/78038025-fad6c080-7374-11ea-8972-708c90a8f2ab.png)

frankyn commented:
@IlyaFaer yes, keep it in the same file to reduce complexity of conformance tests.

frankyn commented:
@IlyaFaer could you update conformance tests to follow the exact format of source conformance tests: https://github.com/googleapis/conformance-tests/blob/master/storage/v1/v4_signatures.json

It can be a subsequent PR to reduce complexity for this one. I'm trying to prevent snowflakes.

frankyn commented on google/cloud/storage/client.py:
```diff
@@ -836,6 +846,175 @@ def get_hmac_key_metadata(
+        bucket_name,
+        blob_name,
+        conditions,
+        expiration,
```
Please put `expiration` before `conditions`.
This should also be changed in example.

IlyaFaer replied:
Oh, yeah, you mentioned that yesterday in fact, I just didn't completely got it, sorry

frankyn commented on google/cloud/storage/client.py:
```diff
@@ -836,6 +846,175 @@ def get_hmac_key_metadata(
+        self,
+        bucket_name,
+        blob_name,
+        conditions,
```
`conditions` parameter should be optional.

IlyaFaer replied:
Done

frankyn commented on tests/unit/test_client.py:
```diff
@@ -1626,8 +1626,8 @@ def test_get_signed_policy_v4_virtual_hosted_style(self):
                 "object-name",
-                conditions=[],
                 expiration=datetime.datetime(2020, 3, 12),
+                conditions=[],
```
`conditions=[]` isn't required for this call, I'd prefer removing empty lists in tests to keep them clean.

IlyaFaer replied:
Done

IlyaFaer commented on google/cloud/storage/client.py:
```diff
@@ -836,6 +846,175 @@ def get_hmac_key_metadata(
+
+        conditions = conditions or []
+        policy_fields = {}
+        if fields:
```
I think, we can write it right in the `for` cycle

IlyaFaer commented:
> follow the exact format of source conformance tests

@frankyn, you mean, use these `signingV4Tests` and `postPolicyV4Tests` fields on a first level of the json? That's not a problem, though I assume some code should be changed in `test__signing.py` as well, so yes, that probably should be done in another PR.

@crwilcox, @frankyn, I think, I saw random conformance tests are flakely failing in Python 2.7 and 3.5. Will take a closer look.

frankyn commented:
Hi @IlyaFaer,
> @frankyn, you mean, use these signingV4Tests and postPolicyV4Tests fields on a first level of the json? That's not a problem, though I assume some code should be changed in test__signing.py as well, so yes, that probably should be done in another PR.

Yes, that's correct, could you please open a tracking issue for it as well? Thank you.

> I think, I saw random conformance tests are flakely failing in Python 2.7 and 3.5. Will take a closer look.

Conformance tests should not have flakes, please keep us posted.

crwilcox commented:
It seems there are a few kokoro failures. @IlyaFaer  are these expected or is this work to do yet?

frankyn commented:
@crwilcox, @IlyaFaer raised that are some flakes. It looks like system tests and not conformance tests, but I might be incorrect. Pending follow-up.

Thanks for reviewing this PR @crwilcox!

IlyaFaer commented:
@frankyn, @crwilcox, I've found out what was causing flaky: `dict.items()` returns result in different order sometimes, so I've added `sorted()`. Didn't detect any failing so far.

I've fixed system tests as well (they were failing with `Anonymous requests are not allowed` for some time, but now they're okay, so I've fixed my bads).

I hope nothing is forgotten. All checks are green now :tada:

crwilcox commented:
Awesome. Yep @IlyaFaer ordered dicts weren't in 2.7 at all or even older versions of 3.
